hibernate-core/src/main/java/org/hibernate/dialect/pagination/Oracle12LimitHandler.java

```diff
@@ -0,0 +1,170 @@
+			switch ( lockMode ) {
+				case UPGRADE:
+				case PESSIMISTIC_READ:
+				case PESSIMISTIC_WRITE:
+				case UPGRADE_NOWAIT:
+				case FORCE:
+				case PESSIMISTIC_FORCE_INCREMENT:
+				case UPGRADE_SKIPLOCKED:
+					return processSql( sql, selection );
+				default:
+					return processSqlOffsetFetch( sql, hasFirstRow );
+			}
+		}
+		return processSqlOffsetFetch( sql, hasFirstRow );
+	}
+
+	private String processSqlOffsetFetch(String sql, boolean hasFirstRow) {
+		bindLimitParametersInReverseOrder = false;
+		final int forUpdateLastIndex = getForUpdateIndex( sql );
+
+		if ( forUpdateLastIndex > -1 ) {
+			return processSql( sql, forUpdateLastIndex, hasFirstRow );
+		}
+
+		sql = sql.trim();
+		final int offsetFetchLength;
+		final String offsetFetchString;
+		if ( hasFirstRow ) {
+			offsetFetchString = " offset ? rows fetch next ? rows only";
+		}
+		else {
+			offsetFetchString = " fetch first ? rows only";
+		}
+		offsetFetchLength = sql.length() + offsetFetchString.length();
+
+		return new StringBuilder( offsetFetchLength ).append( sql ).append( offsetFetchString ).toString();
+	}
+
+	private String processSql(String sql, int forUpdateIndex, boolean hasFirstRow) {
+		bindLimitParametersInReverseOrder = true;
+		sql = sql.trim();
+
+		String forUpdateClause = null;
+		boolean isForUpdate = false;
+		if ( forUpdateIndex > -1 ) {
+			// save 'for update ...' and then remove it
+			forUpdateClause = sql.substring( forUpdateIndex );
+			sql = sql.substring( 0, forUpdateIndex - 1 );
+			isForUpdate = true;
+		}
+
+		final StringBuilder pagingSelect;
+
+		final int forUpdateClauseLength;
+		if ( forUpdateClause == null ) {
+			forUpdateClauseLength = 0;
+		}
+		else {
+			forUpdateClauseLength = forUpdateClause.length() + 1;
+		}
+
+		if ( hasFirstRow ) {
+			pagingSelect = new StringBuilder( sql.length() + forUpdateClauseLength + 98 );
+			pagingSelect.append( "select * from ( select row_.*, rownum rownum_ from ( " );
+			pagingSelect.append( sql );
+			pagingSelect.append( " ) row_ where rownum <= ?) where rownum_ > ?" );
+		}
+		else {
+			pagingSelect = new StringBuilder( sql.length() + forUpdateClauseLength + 37 );
+			pagingSelect.append( "select * from ( " );
+			pagingSelect.append( sql );
+			pagingSelect.append( " ) where rownum <= ?" );
+		}
+
+		if ( isForUpdate ) {
+			pagingSelect.append( " " );
+			pagingSelect.append( forUpdateClause );
+		}
+
+		return pagingSelect.toString();
+	}
+
+	private int getForUpdateIndex(String sql) {
+		final int forUpdateLastIndex = sql.toLowerCase( Locale.ROOT ).lastIndexOf( "for update" );
+		// We need to recognize cases like : select a from t where b = 'for update';
+		final int lastIndexOfQuote = sql.lastIndexOf( "'" );
+		if ( forUpdateLastIndex > -1 ) {
+			if ( lastIndexOfQuote == -1 ) {
+				return forUpdateLastIndex;
+			}
+			if ( lastIndexOfQuote > forUpdateLastIndex ) {
+				return -1;
+			}
+			return forUpdateLastIndex;
+		}
+		return forUpdateLastIndex;
```

Comment:
The above code handles the example case perfectly but is it perfectly enough without ANTLR parsing? IMHO, v6's ANTLR parsing result is the only reliable way to be robust. Maybe this is enough for v5 but when we merge this PR to v6, we might at least leave some TODO comment to remind us to implement it in better way.
Not 100% sure about my remarks, though.

Reply:
not sure I have understood your point, here we are trying to manage the oracle specific sql code needed to the limit ResultSet

Reply:
My point is SQL string tweaking has its limit and we can easily think of bad cases to defeat our simple string processing. For instance, is it possible that 'do update' might appear in some SQL comment in the middle? Just an example but if you check the SQL grammar, it is not easy to cover all the possibilities. That is one of the selling points of v6: we rely on ANTLR or semantic presentation in favour of SQL string tweaking. 
Feel free to ignore my comment, as always.